pep-0484.txt

```diff
@@ -1120,6 +1120,56 @@ subtype of ``Type[Base]``::
       ...
 
 
+Annotating instance and class methods
+-------------------------------------
+
+In most cases the first argument of class and instance methods
+does not need to be annotated, and it is assumed to have the
+type of the containing class for instance methods, and a type object
+type corresponding to the containing class object for class methods.
+In addition, the first argument in an instance method can be annotated
+with a type variable. In this case the return type uses the same
+type variable, thus making that method a generic function. For example::
+
+  T = TypeVar('T', bound='Copyable')
+  class Copyable:
+      def copy(self: T) -> T:
+          # return a copy of self
+
+  class C(Copyable): ...
+  c = C()
+  c2 = c.copy()  # type here should be C
+
+The same applies to class methods using ``Type[]`` in an annotation
+of the first argument::
+
+  T = TypeVar('T', bound='C')
+  class C:
+      def factory(cls: Type[T]) -> T:
+          # make a new instance of cls
+
+  class D(C): ...
+  d = D.factory()  # type here should be D
+
+Note that some type checkers may apply restrictions on this use, such as
+requiring an appropriate upper bound for the type variable used
+(see examples). Also this use case should not be confused with the use
```

Comment:
Yes, for the ``Num`` one it is normal due to invariance.
And also for the second part indeed it worked OK just by chance, after checking all combinations, I have found this one
```python
N = TypeVar('N', bound='Float')
class Float:
    def __init__(self, val) -> None: ...
    def compare(self: N, other: N) -> N: ... # note that I put N in return
class Int(Float): ...

fci = Float(3.14).compare(Int(1))
reveal_type(fci) # Revealed type is 'test4.Int*'
```
But if ``Float.compare`` follows the semantics of normal generic function it should be ``Float``.
That means that this in fact also does not work in mypy, and therefore there is no mystery.
Sorry :-)